draft-js-inline-toolbar-plugin/src/components/Toolbar/index.js

```diff
@@ -45,23 +61,126 @@ export default class Toolbar extends React.Component {
    */
   onOverrideContent = (overrideContent) => {
     this.setState({ overrideContent });
-  }
+  };
 
   onSelectionChanged = () => {
     // need to wait a tick for window.getSelection() to be accurate
     // when focusing editor with already present selection
     setTimeout(() => {
       if (!this.toolbar) return;
+
+      if (typeof this.state.width !== 'number') {
+        this.setState({ width: this.toolbar.offsetWidth });
+      }
+
+      const metrics = {
+        rule: 'left',
+        ruleValue: 0,
+        shift: null,
+        shiftValue: 0
+      };
+      // boundings of the selected text
+      const selectionRect = getVisibleSelectionRect(window);
+      if (!selectionRect) return;
+
       const relativeParent = getRelativeParent(this.toolbar.parentElement);
-      const toolbarHeight = this.toolbar.clientHeight;
+      if (relativeParent) {
+        // avoid false
+        const fontSize = window.getComputedStyle(relativeParent, null).getPropertyValue('font-size');
+        // exit if no selection was made
+        if (selectionRect.left === selectionRect.right && selectionRect.height === parseFloat(fontSize)) {
+          return;
+        }
+      }
       const relativeRect = (relativeParent || document.body).getBoundingClientRect();
-      const selectionRect = getVisibleSelectionRect(window);
 
-      if (!selectionRect) return;
+      // if parent block width is wider than the toolbar, we must forcibly set its
+      // height to prevent unexpected geometry changes
+      if (relativeRect.width > this.toolbar.offsetWidth) {
+        this.toolbar.style.width = `${this.state.width}px`;
+      }
+
+      const toolbarHalfWidth = this.toolbar.offsetWidth / 2;
+      // calculating the middle of the text selection
+      const fromBeginningToMiddle = (selectionRect.left + (selectionRect.width / 2));
+      // calculating the distance from editor left side to the text selection middle
+      const fromParentBeginning = fromBeginningToMiddle - relativeRect.left;
+      // the same but against editor right side
+      const beforeParentEnd = relativeRect.right - fromBeginningToMiddle;
+
+      // the selection is closer to parent beginning than half of the toolbar
+      // +-----------------------------------------------+
+      // |          vv toolbar                           |
+      // | +------------------+                          |
+      // | +------------------+                          |
+      // |                                               |
+      // |  +--+                                         |
+      // |   ^^ selection                                |
+      // +-----------------------------------------------+
+      if (fromParentBeginning < toolbarHalfWidth) {
```

Comment:
rather than the parent, shouldn't it just be the viewport?

Reply:
If I understood you correctly this is also done. Now it looks like this:
![inlinetoolbar](https://user-images.githubusercontent.com/11130487/39228175-ead316c2-4886-11e8-8966-26fc8a08f1fb.png)